File: src/javascript/algorithms/linkedlist/5.removeKthNodeFromEnd/removeKthNodeFromEnd.js
// Using Two pointers:- TC: O(n) SC: O(1)

function removeKthNode(head, k) {
    let dummy = new Node(-1);
    dummy.next = head;
    let first = dummy, second = head;

    while(k >0 && second !== null) {
        second = second.next;
        k--;
    }

    while(second !== null) {
        first = first.next;
        second = second.next;
    }

    first.next = first.next.next;
    return dummy.next;
}

function printLinkedList(head) {
    let temp = head;
    while(temp != null) {
        console.log(temp.value + " ");
        temp = temp.next;
    }
}

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }

    setNext(next) {
        this.next = next;
    }
}

let headNode1 = new Node(1);
let node2 = new Node(2);
let node3 = new Node(3);
let node4 = new Node(4);
let node5 = new Node(5);
let node6 = new Node(6);

let headNode2 = new Node(3);

headNode1.setNext(node2);
node2.setNext(node3);
node3.setNext(node4);
node4.setNext(node5);
node5.setNext(node6);

printLinkedList(removeKthNode(headNode1, 2));
printLinkedList(removeKthNode(headNode2, 1));

File: src/python/algorithms/linkedlist/5.removeKthNodeFromEnd/remove_kth_node_from_end.py
"""
Remove Nth Node From End of List

Given the head of a linked list, remove the nth node from the end and return its head.

Time Complexity: O(n)
Space Complexity: O(1)
"""


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def remove_nth_from_end(head, n):
    """
    Two pointer approach.
    TC: O(n), SC: O(1)
    """
    dummy = ListNode(0, head)
    slow = fast = dummy
    
    # Move fast pointer n+1 steps ahead
    for _ in range(n + 1):
        fast = fast.next
    
    # Move both pointers until fast reaches the end
    while fast:
        slow = slow.next
        fast = fast.next
    
    # Remove the nth node
    slow.next = slow.next.next
    
    return dummy.next


def print_list(head):
    """Helper to print linked list."""
    values = []
    while head:
        values.append(str(head.val))
        head = head.next
    return " -> ".join(values)


# Test case
if __name__ == "__main__":
    # Create list: 1 -> 2 -> 3 -> 4 -> 5
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    
    print(f"Original: {print_list(head)}")
    result = remove_nth_from_end(head, 2)
    print(f"After removing 2nd from end: {print_list(result)}")  # 1 -> 2 -> 3 -> 5
